core/modules/drr.cc

```diff
@@ -0,0 +1,392 @@
+      int err = llring_sc_dequeue(f->queue, reinterpret_cast<void**>(&pkt));
+      if (err < 0)
+        return err;
+    } else {
+      pkt = f->next_packet;
+      f->next_packet = nullptr;
+    }
+
+    if (pkt->total_len() > f->deficit) {
+      f->next_packet = pkt;
+      break;
+    }
+
+    f->deficit -= pkt->total_len();
+    total_bytes += pkt->total_len();
+    batch->add(pkt);
+  }
+
+  return total_bytes;
+}
+
+DRR::Flow_Id DRR::GetId(bess::Packet* pkt) {
+  struct ether_hdr *eth = pkt->head_data<struct ether_hdr *>();
+  struct ipv4_hdr *ip = reinterpret_cast<struct ipv4_hdr *>(eth + 1);
+  int ip_bytes = (ip->version_ihl & 0xf) << 2;
+  struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(
+      reinterpret_cast<uint8_t *>(ip) + ip_bytes);  // Assumes a l-4 header
+  // TODO(joshua): handle packet fragmentation
+  Flow_Id id = {ip->src_addr, ip->dst_addr, udp->src_port,
+      udp->dst_port, ip->next_proto_id};
+  return id;
+}
+
+int DRR::AddNewFlow(bess::Packet *pkt, Flow_Id id) {
+  int err;
+
+  // creates flow
+  Flow* f = static_cast<Flow *>(mem_alloc_ex(sizeof(Flow), alignof(Flow), 0));
+  Flow init = {0, 0, id, 0, nullptr};
+  memcpy(f, &init, sizeof(init));
+
+  // TODO(joshua) do proper error checking
+  err = AddQueue(&f->queue, static_cast<int>(kFLOW_QUEUE_SIZE));
+
+  if (err != 0)
+    return err;
+
+  flows_.Insert(id, f);
+
+  err = Enqueue(f, pkt);
+  if (err != 0)
+    return err;
+
+  // puts flow in round robin
+  err = llring_sp_enqueue(flow_ring_, f);
+  if (err != 0)
+    return err;
+
+  return 0;
+}
+
+void DRR::RemoveFlow(Flow* f) {
+  if (f == current_flow_) {
+    current_flow_ = nullptr;
+  }
+
+  bess::Packet* pkt;
+  while (llring_sc_dequeue(f->queue, reinterpret_cast<void**>(&pkt)) == 0) {
+    bess::Packet::Free(pkt);
+  }
+
+  if (f->next_packet) {
+    bess::Packet::Free(f->next_packet);
+  }
+
+  mem_free(f->queue);
+  flows_.Remove(f->id);
+  mem_free(f);
+}
+
+int DRR::AddQueue(llring** queue, int slots) {
+  int bytes = llring_bytes_with_slots(slots);
+  int ret;
+
+  *queue = static_cast<llring *>(mem_alloc_ex(bytes, alignof(llring), 0));
+  if (!*queue) {
+    return -ENOMEM;
+  }
+
+  ret = llring_init(*queue, slots, 1, 1);
+  if (ret) {
+    mem_free(*queue);
+    return -EINVAL;
+  }
+  return 0;
+}
+
+int DRR::Enqueue(Flow* f, bess::Packet* newpkt) {
+  int ret;
+
+  // creates a new queue if there is not enough space for the new packet
+  // in the old queue
+  if (llring_full(f->queue)) {
+    if (llring_count(f->queue)*kQUEUE_GROWTH_FACTOR > max_queue_size_) {
+      bess::Packet::Free(newpkt);
+      return 0;
+    }
+    int slots = RoundToPowerTwo(llring_count(f->queue)*kQUEUE_GROWTH_FACTOR);
+    ret = ResizeQueue(&f->queue, slots);
+    if (ret != 0) {
+      return ret;
+    }
+  }
+  ret = llring_sp_enqueue(f->queue, reinterpret_cast<void*>(newpkt));
+  f->timer = get_epoch_time();
+  return ret;
+}
+
+int DRR::ResizeQueue(llring **queue_loc, int new_size) {
```

Comment:
Here too it would be better to return an llring * and have an int *error second argument.  Then the semantics of the method would change slightly, in that the method would be destructive to its first argument (which would be an llring *old_queue).